src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Logging/RoslynLogger.cs

```diff
@@ -230,28 +231,38 @@ private static LogType GetKind(LogMessage logMessage)
                                         _ => LogType.Trace
                                     };
 
-        private static ImmutableDictionary<string, object?> GetProperties(FunctionId id, LogMessage logMessage, int? delta)
+        private static ReadOnlyMemory<Property> GetProperties(FunctionId id, LogMessage logMessage, int? delta)
         {
-            var builder = ImmutableDictionary.CreateBuilder<string, object?>();
+            using var pooledObject = s_propertyArrayPool.GetPooledObject();
+
+            var properties = pooledObject.Object;
+            var propCount = 0;
 
             if (logMessage is KeyValueLogMessage kvLogMessage)
             {
+                // If the given properties array isn't large enough, allocate one
+                if (kvLogMessage.Properties.Count + 1 >= properties.Length)
+                    properties = new Property[properties.Length + 1];
+
                 foreach (var (name, val) in kvLogMessage.Properties)
                 {
-                    builder.Add(GetPropertyName(id, name), val);
+                    properties[propCount] = new(GetPropertyName(id, name), val);
+                    propCount++;
                 }
             }
             else
             {
-                builder.Add(GetPropertyName(id, "Message"), logMessage.GetMessage());
+                properties[propCount] = new(GetPropertyName(id, "Message"), logMessage.GetMessage());
+                propCount++;
             }
 
             if (delta.HasValue)
             {
-                builder.Add(GetPropertyName(id, "Delta"), delta.Value);
+                properties[propCount] = new(GetPropertyName(id, "Delta"), delta.Value);
+                propCount++;
             }
 
-            return builder.ToImmutableDictionary();
+            return new ReadOnlyMemory<Property>(properties, 0, propCount);
```

Comment:
no :)

Reply:
Not the first time I've caused an issue like this during a refactoring.